docs/pattern-matching.md

```diff
@@ -1,229 +1,306 @@
+Declarations can combine patterns with `|` if the same variables are declared in each pattern.
 
 ```reason
-let optionBoolToBool = opt =>
-  switch (opt) {
-  | Some(trueOrFalse) => trueOrFalse
-  | None => false
-  };
+type t = A(int) | B(int);
+let data = B(3);
+let A(v) | B(v) = data; 
+/* v is 3 */
 ```
 
-Pretty darn hard to make a mistake in this code at this point! Whenever you'd like to use an if-else with many branches, prefer pattern matching instead. It's more concise and [performant](variant.md#design-decisions) too.
+## Other pattern matching features
 
-See another example, with switch + tuple [here](tuple.md#tips-tricks).
+### when
 
-## Design Decisions
+`when` can add extra conditions to patterns. The condition must be satisfied in order to execute the pattern's code, otherwise the pattern is skipped. Note that `when` should be used carefully since the exhaustive and unused pattern warnings do not analyze their conditions.
 
-The notorious [fizzbuzz problem](https://en.wikipedia.org/wiki/Fizz_buzz#Programming_interviews) strangely trips up some people, partially due its nature of paralyzing the programmer who hopes to simplify/unify the few condition branches in search of elegance where there's none. Hopefully you can see that usually, pattern-matching's visual conciseness [allows us to overcome decision paralysis](/try.html?reason=PQKgBAQghgzgpgEzAewHZgBYBcsAcYBcwwATsvDlAMbIJwB0yJA5sAO4CWA1h8AGIcAXoIgBXYQGIA8gGEoAWwA2YEMABQiuFjAAzIYIBG4wWAC8YABQcAlGYB8asGBicsVDJY5h5tMAGYAGjAvHyQAVlsAb0cwAB9LAAYghNtTOzAAIgFhMWEMmPiLJLAAfVT0rP18p0KS5PLM3MFquNL7ZywSDlRmEuQdEu6sK2sYgF8AbjU1HSZPYPQARjAsZDBFhJSwaKcAKRh6RWRmCz1hI2ER0cmgA), while keeping all the benefits (and more, as you've seen) of a bunch of brute-forced `if-else`s. There's really nothing wrong with explicitly listing out all the possibilities; Pattern matching corresponds to **case analysis** in math, a valid problem-solving technique that proves to be extremely convenient.
+```reason
+let k =
+  switch (p) {
+  | {x, y: 0} when x * x <= 4
+  | {x, y: 0} when f(x)
+  | {x: 2, y} when y < 10 => 1
+  | {x: 2, y} when y < 2 => 2 /* never executed, but no warning */
+  | _ => 3
+  };
+/* k is 1 */
+```
 
-Using a Reason `switch` for the first time might make you feel like you've been missing out all these years. Careful, for it might ruin other languages for you =).
+### Matching exceptions
 
-If you've tried to refactor a big, nested if-else logic, you might realize it's very hard to get the logic right. On the other hand, pattern matching + tuple conceptually maps to a 2D table, where each cell can be independently filled. This ensures that whenever you need to add a case in the `switch`, you can target that and only that table cell, without messing other cells up.
+If executing an input raises an exception, it can be matched alongside the other values of the input:
```

Comment:
Does it make sense to add an example of `try/catch` pattern matching in isolation here too, not just exceptions used in a switch. This gives the impression that this is the only way to handle exceptions to me